nengo/params.py

```diff
@@ -49,10 +49,18 @@ def iter_params(obj):
 
 def equal(a, b):
     """Check if two (possibly array-like) objects are equal."""
+    if isinstance(a, dict) and isinstance(b, dict):
+        if set(a) != set(b):
```

Comment:
`return set(a) == set(b) and all(equal(a[key], b[key]) for key in a)`

Reply:
I think I did this for efficiency reasons and possible typeerrors if `a` or `b` aren't dicts, but I'll look again.

Reply:
It should be just as efficient and safe the way I have it, because both `and` and `all` are lazy (as long as you use a generator in `all` as above).

Reply:
It's the conversion to `set` that I'm aiming to avoid, can be costly for big dicts. But it's probably a premature optimization.

Reply:
I don't understand how your code avoids a conversion to `set`. It seems to me that both `a` and `b ` get converted to set in this first line.

Reply:
Could do `all(k in b for k in a) and all(k in a for k in b)` instead of the `set(a) == set(b)` to avoid converting to `set`.

Reply:
> It seems to me that both `a` and `b ` get converted to set in this first line.

Not if either a or b is not a dict. E.g. if one is a tuple or list or really long numpy array (the most likley scenario in which it could be bad performance-wise). Anyway, I'll look at this again, probably it can be simplified to one statement.

Reply:
The `if isinstance(a, dict) and isinstance(b, dict)` above this line means both `a` and `b` will be dicts though? But if you were to get rid of that guard, converting to `set` may be generally cheaper than `k in a for k in b` or vice-versa as the latter is `O(n^2)` worst case for two arrays whereas `set(a) == set(b)` is `O(n)`.

Reply:
Yes, I assumed we were keeping the guard (i.e. we know they're both dicts). I don't see how this would work without it. My original comment was just a way to condense the same logic as Trevor originally had into one line (two if you count the guard) so that it's more readable. There shouldn't be any change to the logic.

Reply:
i.e. instead of
```
    if isinstance(a, dict) and isinstance(b, dict):
        if set(a) != set(b):
            return False
        for key in a:
            if not equal(a[key], b[key]):
                return False
        return True
```
we have
```
    if isinstance(a, dict) and isinstance(b, dict):
        return set(a) == set(b) and all(equal(a[key], b[key]) for key in a)
```